src/mutator.rs

```diff
@@ -0,0 +1,282 @@
+//! Mutator: an entity that changes shared locations
+//!
+//! # Registration
+//!
+//! In order to track all mutators in one place, we need some form of mutator registration. When a
+//! mutator is created, it is registered to a global lock-free singly-linked list of registries; and
+//! when a mutator is dropped, it is unregistered from the list.
+//!
+//! # Pinning
+//!
+//! Every registry contains an integer that tells whether the mutator is pinned and if so, what was
+//! the global epoch at the time it was pinned. Mutators also hold a pin counter that aids in
+//! periodic global epoch advancement.
+//!
+//! When a mutator is pinned, a `Scope` is returned as a witness that the mutator is pinned.  Scopes
+//! are necessary for performing atomic operations, and for freeing/dropping locations.
+
+use std::cell::Cell;
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering::{Relaxed, Release, SeqCst};
+
+use atomic::Ptr;
+use sync::list::Node;
+use global;
+
+
+/// Number of pinnings after which a mutator will collect some global garbage.
+const PINS_BETWEEN_COLLECT: usize = 128;
+
+
+/// Entity that changes shared locations.
+pub struct Mutator<'scope> {
+    /// This mutator's entry in the registry list.
+    registry: &'scope Node<Registry>,
+    /// Whether the mutator is currently pinned.
+    is_pinned: Cell<bool>,
+    /// Total number of pinnings performed.
+    pin_count: Cell<usize>,
+}
+
+/// An entry in the linked list of the registered mutators.
+#[derive(Default, Debug)]
+pub struct Registry {
```

Comment:
This is not a registry -- the registry is the linked list that holds thread-local epochs, and this struct is just an entry in the registry. Can we name this struct differently?